pkg/rootless/cni/rootless_cni.go

```diff
@@ -0,0 +1,349 @@
+	// this also has to be a user writable location
+	RootlessCNIInfraCmd         = "/run/rootless-cni-infra-exe"
+	rootlessCNIInfraCreateNetNS = "rootless-cni-infra-create-netns"
+	basePath                    = "/run/rootless-cni-infra"
+)
+
+// Config passed to alloc via stdin
+type Config struct {
+	PluginPaths    []string
+	NetConfPath    string
+	IfName         string
+	Args           [][2]string
+	CapabilityArgs map[string]interface{} `json:"CapabilityArgs,omitempty"`
+}
+
+// AllocResult returned by alloc as json
+type AllocResult struct {
+	CNIResult cnitypes.Result `json:"cni_result"`
+	NSPath    string          `json:"ns_path"`
+}
+
+// IsIdle returned by is-idle as json
+type IsIdle struct {
+	Idle bool `json:"idle"`
+}
+
+func printError(format string, a ...interface{}) {
+	fmt.Fprintf(os.Stderr, "Error: "+format+"\n", a...)
+}
+
+func printJSONResult(v interface{}) {
+	b, err := json.Marshal(v)
+	if err != nil {
+		printError("%s", err)
+	}
+	fmt.Println(string(b))
+}
+
+func init() {
+	reexec.Register(RootlessCNIInfraCmd, func() {
+		if len(os.Args) < 2 {
+			printError("%s requires at least one arg", RootlessCNIInfraCmd)
+			os.Exit(1)
+		}
+
+		switch os.Args[1] {
+		case "alloc":
+			if len(os.Args) != 4 {
+				printError("%s alloc requires exactly 2 args", RootlessCNIInfraCmd)
+				os.Exit(1)
+			}
+			alloc(os.Args[2], os.Args[3])
+
+		case "dealloc":
+			if len(os.Args) != 4 {
+				printError("%s dealloc requires exactly 2 args", RootlessCNIInfraCmd)
+				os.Exit(1)
+			}
+			dealloc(os.Args[2], os.Args[3])
+
+		case "is-idle":
+			idle := IsIdle{
+				Idle: false,
+			}
+			empty, err := dirIsEmpty(basePath)
+			if os.IsNotExist(err) || empty {
+				idle.Idle = true
+			} else if err != nil {
+				printError("%s", err)
+			}
+			printJSONResult(idle)
+
+		case "sleep":
+			// sleep subcommand used to keep the namespace alive
+			// sleep max duration
+			time.Sleep(time.Duration(1<<63 - 1))
+
+		default:
+			printError("Unknown command: %s %s", RootlessCNIInfraCmd, os.Args[1])
+			os.Exit(1)
+		}
+	})
+
+	reexec.Register(rootlessCNIInfraCreateNetNS, func() {
+		if len(os.Args) != 2 {
+			printError("%s requires one arg", rootlessCNIInfraCreateNetNS)
+			os.Exit(1)
+		}
+		pidfile := os.Args[1]
+		if err := os.MkdirAll(path.Dir(pidfile), 0700); err != nil {
+			exit(err)
+		}
+		// create new net namespace
+		if err := syscall.Unshare(syscall.CLONE_NEWNET); err != nil {
+			exit(err)
+		}
+
+		// background process to keep the net namespace alive
+		sleep := reexec.Command(RootlessCNIInfraCmd, "sleep")
+		if err := sleep.Start(); err != nil {
+			exit(err)
+		}
+		pid := sleep.Process.Pid
+		stringPid := strconv.Itoa(pid)
+
+		if err := ioutil.WriteFile(pidfile, []byte(stringPid), 0700); err != nil {
+			exit(errors.Wrap(err, "failed to write pid file"))
+		}
+
+		// set the loopback adapter up
+		cmd := exec.Command("ip", "link", "set", "lo", "up")
+		if err := cmd.Run(); err != nil {
+			exit(err)
```

Comment:
should print error from `ip` command

Reply:
(or use netlink without shelling out `ip`)

Reply:
I changed this to use netlink.